File: 152.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        if(nums.size() == 0) return 0;
        int curMax = nums[0], curMin = nums[0], maxPro = nums[0];
        for (int i=1; i<nums.size(); i++){
            int tmp = curMax * nums[i];
            curMax = max(nums[i], max(curMax*nums[i], curMin*nums[i]));
            curMin = min(nums[i], min(tmp, curMin*nums[i]));
            maxPro = max(maxPro, curMax);
        }
        return maxPro;
    }
};

File: go-solutions/152.go
package main

func maxProduct(nums []int) int {
	if len(nums) < 1 {
		return 0
	}

	dp := make([][2]int, len(nums))
	dp[0] = [2]int{
		nums[0], nums[0],
	}

	res := nums[0]
	for i := 1; i < len(nums); i++ {
		dp[i][0] = max(max(dp[i-1][0]*nums[i], dp[i-1][1]*nums[i]), nums[i])
		dp[i][1] = min(min(dp[i-1][0]*nums[i], dp[i-1][1]*nums[i]), nums[i])
		res = max(res, dp[i][0])
	}
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
